### PR TITLE
fix(core): Preserve branch output index on agent resume to maintain `pairedItem` chain

### LINKED ISSUE #26592: Expression `.item` syntax fails to resolve in UI after branching node when AI Agent executes tool calls
### Bug Description

When an AI Agent node is placed after a branching node (If/Switch) and **actually executes tool calls**, expression references using `.item` syntax to access nodes before the branching node become undefined/unresolvable in the UI parameter panel.

**Key symptoms:**
- Expressions show as RED/undefined in the parameter panel after execution
- The same workflow executes successfully (data is accessible during execution)
- The Input section shows the data correctly
- Using `.first()` instead of `.item` works as a workaround
- The bug does NOT occur if the Agent doesn't call any tools

**Impact:** Users cannot use `.item` syntax in AI Agent prompts to reference context from earlier nodes when there's a branching node in between.

### To Reproduce

**Minimal workflow:**
```
Manual Trigger → If → AI Agent + Calculator Tool
```

**Steps:**

1. Create **Manual Trigger** with data:
   ```json
   [{ "number": 10, "name": "test_context" }]
   ```

2. Add **If** node:
   - Condition: `{{ $json.number == 9 }}`

3. Add **AI Agent** node (connect from false branch):
   - Connect any Chat Model (OpenAI, Anthropic, etc.) -- **NOTE**: In my env I was forced to use a proprietary LLM gateway for which I wrote a custom chat node. I _suspect_ (but I cannot confirm) that this will be the same with any chat model.  
   - Connect **Calculator Tool**
   - Prompt (forces tool usage):
     ```
     Calculate ($('Manual Trigger').item.json.number + 5) * 3 using the calculator tool.
     ```

4. Execute workflow (agent calls calculator successfully)

5. Click on AI Agent node after execution

**Observed:**
- Expression `{{ $('Manual Trigger').item.json.number }}` shows RED/undefined
- Dragging new reference from Input section briefly shows green, then turns red

**Key finding:** If you change the prompt to NOT use the calculator (e.g., "Just say hello"), the expression stays GREEN. The bug only triggers when the agent **actually executes a tool call**.

**Tested with:**
- Calculator tool (built-in) ✅ Bug reproduces
- MCP Client tool ✅ Bug reproduces
- Agent without tool calls ✅ No bug (works fine)
- No If node (just Manual Trigger → Agent) ✅ No bug (works fine)

### Expected behavior

After workflow execution:
- Expressions using `.item` syntax should remain green and resolvable
- The UI should correctly trace the paired item chain: `Agent item → If branch → Manual Trigger item`
- This should work regardless of whether the Agent executed tool calls

**Workaround:** Replace `.item` with `.first()`:
```javascript
{{ $('Manual Trigger').first().json.number }}  // Works but breaks paired item semantics
```

### Debug Info

# Debug info

## core

- n8nVersion: 2.9.2
- platform: npm
- nodeJsVersion: 22.21.1
- nodeEnv: undefined
- database: sqlite
- executionMode: regular
- concurrency: -1
- license: community
- consumerId: unknown

## storage

- success: all
- error: all
- progress: false
- manual: true
- binaryMode: filesystem

## pruning

- enabled: true
- maxAge: 336 hours
- maxCount: 10000 executions

## client

- userAgent: mozilla/5.0 (windows nt 10.0; win64; x64; rv:149.0) gecko/20100101 firefox/149.0
- isTouchDevice: false

Generated at: 2026-03-05T10:35:04.972Z

### Operating System

Windows WSL -  Ubuntu 24.04.3 LTS

### n8n Version

2.9.2

### Node.js Version

v22.21.1

### Database

SQLite (default)

### Execution mode

main (default)

### Hosting

self hosted

### LINKED ISSUE #27767: [Bug]: AI Agent (v3.1) throws $getPairedItem lineage error when connected to an If node (observed on 'false' branch), resulting in empty success output
### Bug Description

I am experiencing a critical and intermittent issue where the AI Agent node (version 3.1) throws a $getPairedItem error when it is connected directly to an If node (specifically observed on the false output / Output Index 1).

Important Context:

While the AI Agent and its connected tools do fetch data from nodes that executed prior to the If node, absolutely no expressions reference the If node itself.

The behavior is highly intermittent. It can happen whether the underlying tool execution succeeds perfectly or fails.

Tool Dependency Unknown: In our specific setup, the AI Agent has several tools connected and is actively using them when the error occurs. However, we have not tested if the error still triggers on a vanilla AI Agent without any tools connected.

Branch Dependency Unknown: We encountered this specifically when the agent is connected to the false branch. We have not tested if connecting it to the true branch yields the exact same error. This might be a general lineage resolution failure between the If node and the AI Agent, rather than strictly tied to the false branch.

The most critical issue: Whenever this lineage error triggers, the AI Agent outputs absolutely nothing (empty text). However, instead of triggering the 'Error' output branch, it routes this empty response through the 'Success' branch. This causes subsequent downstream nodes (like HTTP requests expecting a chat message string) to crash.

Actual Behavior:
Intermittently, when the AI Agent executes, n8n throws the following lineage error behind the scenes:

Paired item data for $getPairedItem from node 'If1' is unavailable. Ensure 'If1' is providing the required output.
An expression here won't work because it uses .item and n8n can't figure out the matching item. The node 'If1' didn't return enough information.

Because of how the AI Agent handles internal state, it finishes its run but outputs an empty string. It then sends this empty output down the Success path. Downstream nodes fail because the expected generated text from the LLM is missing.

Workaround (Mitigation):
Placing a No Operation, do nothing (NoOp) node directly between the If node and the AI Agent node completely mitigates the issue. The AI Agent reads from Output Index 0 of the NoOp node, easily traces the lineage back to previous nodes, and the intermittent $getPairedItem errors stop happening.

### To Reproduce

1. Create a workflow where data flows through several nodes, then into an If node.

2. Evaluate a condition that routes data to the false output of the If node.Connect an AI Agent node (v3.1) directly to this false output.

3. Attach one or more tools to the AI Agent. Inside the tools or the agent's prompt, fetch data from nodes that executed before the If node.

4. Provide a prompt that forces the AI Agent to process the input (and call a tool, if attached).

5. Run this multiple times to observe the intermittent failure.

### Expected behavior

The AI Agent should be able to resolve its execution context and maintain lineage to nodes prior to the If node, regardless of whether it is connected to an If node directly. Furthermore, if a critical lineage error like this occurs, the AI Agent should route the execution to the Error branch, not the Success branch with empty data.

### Debug Info

# Debug info

## core

- n8nVersion: 2.10.2
- platform: docker (self-hosted)
- nodeJsVersion: 24.13.1
- nodeEnv: production
- database: postgres
- executionMode: scaling (single-main)
- concurrency: -1
- license: community
- consumerId: unknown

## storage

- success: all
- error: all
- progress: false
- manual: true
- binaryMode: database

## pruning

- enabled: true
- maxAge: 336 hours
- maxCount: 10000 executions

## client

- userAgent: mozilla/5.0 (windows nt 10.0; win64; x64) applewebkit/537.36 (khtml, like gecko) chrome/146.0.0.0 safari/537.36
- isTouchDevice: false

Generated at: 2026-03-30T14:14:14.122Z

### Operating System

Ubuntu 24.04

### n8n Version

2.10.2

### Node.js Version

24.13.1

### Database

PostgreSQL

### Execution mode

main (default)

### Hosting

self hosted

### PR DESCRIPTION
**Pull request by sinehypernova-0718**
## Summary

Fix `.item` expression resolution after an AI Agent executes a tool following a branching node (If/Switch). The engine now correctly carries forward the real upstream branch index and the original paired‑item index when resuming the agent, so both the agent’s prompt and downstream nodes can trace back to the triggering item.

While this PR only addresses the AI Agent tool‑call resume path, it also highlights a general weakness in how `pairedItem` metadata is propagated across multiple output branches. The approach used here (preserving `previousNodeOutput`, carrying per‑item original indices, and always attaching `sourceOverwrite`) can guide a similar fix for error‑output scenarios like as in Issue [#30050](<https://github.com/n8n-io/n8n/issues/30050>).

### Root cause

When an agent pauses to call a tool, the execution engine later re-queues the agent node via `handleRequest`. During this resume phase, three things were going wrong:

1. **Wrong source lookup index** – `prepareRequestedNodesForExecution` read `executionData.source?.main?.[runIndex]`, but for agents the synthetic input is always at index `0`. Using the run index caused the engine to pick the wrong branch provenance.
2. **Hard-coded** `parentOutputIndex = 0` – the re-queued agent entry always claimed it came from output `0`, so even if the agent was wired to branch `1` (false branch), downstream nodes thought it came from branch `0`.
3. **Shallow** `pairedItem` **override in** `finalizeResult.ts` – ToolsAgent V3 was replacing any carefully reconstructed `pairedItem` with a bare `{ item: itemIndex }`, stripping the `input`, `sourceOverwrite` and full ancestry that the engine had just built.
4. **Incorrect** `runIndex` **inheritance** – the resumed agent entry inherited the parent’s `previousNodeRun` as its own `runIndex`, which confused source tracking.
5. **Missing batch support** – the original fix only stored a single `originalPairedItemIndex`; multi-item agent batches need an array so every output item can point back to its correct upstream counterpart.

### What changed

| File | Change |
| -- | -- |
| `packages/workflow/src/interfaces.ts` | Added `originalPairedItemIndex?: number` and `originalPairedItemIndices?: number[]` to `ITaskMetadata`. |
| `packages/core/src/execution-engine/requests-response.ts` | • Reads source data from `main[0]` instead of `main[runIndex]`.  <br>• Sets `parentOutputIndex` to the real `previousNodeOutput` (preserves branch index).  <br>• Always emits `preserveSourceOverwrite` / `preservedSourceOverwrite` on resume so the branch node stays traversable.  <br>• Captures **per-item** `originalPairedItemIndices` from the agent’s input before the tool round-trip.  <br>• Keeps the agent’s own `runIndex` instead of overwriting it with the parent’s run index. |
| `packages/core/src/execution-engine/workflow-execute.ts` | • Introduces `sourceDataIndex`: for resumed agents / tool executions the synthetic data lives at index `0`, but `outputIndex` still carries the real branch index for provenance.  <br>• Reconstructs `taskSource` explicitly so `previousNodeOutput` is preserved in `taskStartedData.source`.  <br>• When mapping input items, uses `originalPairedItemIndices[itemIndex]` (or the legacy scalar fallback) for `pairedItem.item`, and `previousNodeOutput` for `pairedItem.input`. |
| `packages/@n8n/nodes-langchain/.../finalizeResult.ts` | **Removes** the shallow `pairedItem: { item: itemIndex }` property. The execution engine now builds the full ancestry object; the helper should not override it. |
| Test files | Added / updated unit tests covering:  <br>• Branch output index > 0  <br>• Missing source fallback  <br>• Correct `runIndex`, `parentOutputIndex`, `preserveSourceOverwrite`  <br>• Single and multi-item `originalPairedItemIndices` propagation  <br>• End-to-end `pairedItem.input` / `pairedItem.item` reconstruction in `workflow-execute.test.ts` |

**How to test:**

1. Import the [provided workflow](<https://github.com/user-attachments/files/27496194/Workflow.json>) (Manual Trigger → If false branch → AI Agent + Calculator → Code).
2. Execute – the Calculator returns the correct result and the Code node’s `.first()` works.
3. Change the Code node to use `$('Manual Trigger').item.json.code` – it now resolves without errors, proving the `pairedItem` chain is intact.

### Backwards compatibility

* The scalar `originalPairedItemIndex` is kept as a fallback in `workflow-execute.ts` for any edge-case metadata that still carries it, but all new resume paths write the array `originalPairedItemIndices`.
* Existing non-agent executions are untouched; the `sourceDataIndex` logic only activates when `nodeWasResumed` or `preserveSourceOverwrite` is set.

## Findings & Broader Impact

While investigating this bug, I discovered that a **complete fix** would require deeper architectural changes beyond this PR:

1. `finalizeResult.ts` – This PR already removes the shallow `pairedItem: { item: itemIndex }` copy, allowing the engine's `assignPairedItems` to build the full ancestry chain from corrected input items.
2. `execute.ts` **(ToolsAgent V2)** – Must preserve full `pairedItem` structure instead of single index (already reverted).
3. `requests-response.ts` – `preservedSourceOverwrite` unconditionally replaces nested provenance (agent‑as‑tool calls when an agent pauses again to call another tool). Should preserve existing value instead of overwriting.
4. **Tests** – Missing coverage for branches, parallel calls, no‑tool edge cases, and nested agent chains.

### Why not included

These would make the PR unreviewable (touching multiple engine files, risk of regressions).

> \[!NOTE\]<br>I am happy to work on those broader changes if the maintainers/reviewers @mutdmour deem them necessary. Please let me know whether you’d like a follow‑up PR that fully rewires `pairedItem` propagation, fixes nested `preservedSourceOverwrite`, and adds the missing tests.

---

## Related Linear tickets, Github issues, and Community forum posts

* fixes [#26592](<https://github.com/n8n-io/n8n/issues/26592>)
* Linear Ticket ID: [CAT-2687](https://linear.app/n8n/issue/CAT-2687)
* fixes [#27767](<https://github.com/n8n-io/n8n/issues/27767>)
* Linear Ticket ID: [CAT-2814](https://linear.app/n8n/issue/CAT-2814)
* related: [#30050](<https://github.com/n8n-io/n8n/issues/30050>) (Error output `pairedItem` break) – not fixed by this PR, but the root‑cause analysis and the engine‑level metadata corrections here could serve as a template for that issue.
* Linear Ticket ID: [CAT-3038](https://linear.app/n8n/issue/CAT-3038)

---

## Review / Merge checklist

- [X] I have seen this code, I have run this code, and I take responsibility for this code.
- [X] PR title and summary are descriptive. ([conventions](<../blob/master/.github/pull_request_title_conventions.md>))
- [ ] [Docs updated](<https://github.com/n8n-io/n8n-docs>) or follow-up ticket created.
- [X] Tests included.
- [ ] PR Labeled with `Backport to Beta`, `Backport to Stable`, or `Backport to v1` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

**Comment by sinehypernova-0718** [comment]
@codex review

File: packages/core/src/execution-engine/requests-response.ts [hunk 1]
```diff
@@ -268,15 +272,30 @@ export function handleRequest({
 		return { nodesToBeExecuted: [] };
 	}
 
+	// Preserve the original pairedItem.item from the agent's input so the resumed
+	// execution can keep the same upstream item lineage after the tool call returns.
+	const originalPairedItem = executionData.data.main?.[0]?.[0]?.pairedItem;
```
[the remark this answers is not in the document]
Comment (sinehypernova-0718) [reply packages/core/src/execution-engine/requests-response.ts]: @codex review

[the remark this answers is not in the document]
Comment (sinehypernova-0718) [reply packages/core/src/execution-engine/requests-response.ts]: @codex address that feedback

File: packages/core/src/execution-engine/requests-response.ts [hunk 2]
```diff
@@ -268,15 +272,31 @@ export function handleRequest({
...
+
 	// 3. add current node back to the bottom of the stack
 	nodesToBeExecuted.unshift({
 		inputConnectionData: result.connectionData,
-		parentOutputIndex: 0,
+		parentOutputIndex: executionData.source?.main?.[0]?.previousNodeOutput ?? 0,
 		parentNode: result.parentNode,
 		parentOutputData: executionData.data.main as INodeExecutionData[][],
-		runIndex,
+		runIndex: executionData.source?.main?.[0]?.previousNodeRun ?? runIndex,
```
[the remark this answers is not in the document]
Comment (sinehypernova-0718) [reply packages/core/src/execution-engine/requests-response.ts]: @codex review

**Comment by sinehypernova-0718** [comment]
Hii! @DeveloperTheExplorer Thanks for the running the CI checks, I see two CI failures. The one in `workflow-execute-process-process-run-execution-data.test.ts` is caused by my `handleRequest` reorder. I'm pushing a fix to make the test expect `runData[tool1Node.name]` to be `undefined` instead of length 1. 
The binary data timeout seems pre‑existing and unrelated to this PR.

Thanks!

**Comment by sinehypernova-0718** [comment]
Hey @DeveloperTheExplorer Really sorry for the delay. we had a network outage here. The test fix is now pushed. Could you please re-trigger the CI when you get a chance? 
Thanks for your patience.

**Comment by sovietspaceship** [comment]
Hi! Sorry this is taking so long to review. I think the best approach would be to open this as an issue so we can look into it from scratch and routed to the correct team. Thanks!
